src/util/teiRelationshipsParser.js

```diff
@@ -113,42 +126,83 @@ export const getDataWithRelationships = async (
         return []
     }
 
-    const isRecursive = from.trackedEntityType.id === to.trackedEntityType.id
+    const isRecursiveTrackedEntityType =
+        from.trackedEntityType.id === to.trackedEntityType.id
+    const isRecursiveProgram =
+        'program' in from &&
+        'program' in to &&
+        from?.program?.id === to?.program?.id
+    const isToProgramDefined = 'program' in to
+
+    // Use target as source if from/to TE Types and Programs match, otherwise
+    // fetch/re-fetch using program if available TE type otherwise
+    let recursiveProp = null
+    if (
+        isRecursiveTrackedEntityType && // Same TE Type
+        !isRecursiveProgram && // Different Program
+        isToProgramDefined // Defined Program
+    ) {
+        recursiveProp = {
+            program: to.program.id,
+        }
+    } else if (
+        isRecursiveTrackedEntityType && // Same TE Type
+        !isRecursiveProgram && // Different Program
+        !isToProgramDefined // Not Defined Program
+    ) {
+        recursiveProp = {
+            type: to.trackedEntityType,
+        }
+    } else if (
+        !isRecursiveTrackedEntityType && // Different TE Type
+        !isRecursiveProgram // Different Program
+    ) {
+        recursiveProp = {
+            type: to.trackedEntityType,
+        }
+    }
 
+    // Keep TEI with coords and simplify structure
     const targetInstances = normalizeInstances(
-        isRecursive
+        isRecursiveTrackedEntityType & isRecursiveProgram
             ? sourceInstances
             : await fetchTEIs({
-                  type: to.trackedEntityType,
+                  ...recursiveProp,
                   fields,
                   orgUnits,
                   organisationUnitSelectionMode,
               })
     )
 
-    const filteredSourceInstances = isRecursive
+    const targetInstanceIds = []
+    // Keep TEI with relationship of correct type
+    // Store Ids of target relationships
+    const filteredSourceInstances = isRecursiveTrackedEntityType
         ? sourceInstances.filter((instance) =>
-              isIndexInstance(instance, relationshipType.id)
+              isIndexInstance(instance, relationshipType.id, targetInstanceIds)
           )
         : sourceInstances
-    const relationshipsById = {}
 
+    const relationshipsById = {}
+    // Create relationship objects
     filteredSourceInstances.forEach((instance) =>
         getInstanceRelationships(
             relationshipsById,
             instance,
             targetInstances,
             relationshipType.id,
-            isRecursive
+            isRecursiveTrackedEntityType
         )
     )
 
-    filteredSourceInstances.forEach((instance) => {
-        delete targetInstances[instance.id]
-    })
+    // Keep only instances that are the target of a relationship
+    const filteredTargetInstances = Object.values(targetInstances).filter(
```

Comment:
Maybe instead of a comment, this var could be renamed to something more informative, like `instancesTargetedByRelationship`.  The current name does not really provide any useful meaning.

Reply:
You might have a better suggestion for the name.

Reply:
change `targetInstances` to `potentialTargetInstances` and `filteredTargetInstances` to `targetInstances`.